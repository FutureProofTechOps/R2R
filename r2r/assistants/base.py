import json
from abc import ABCMeta, abstractmethod
from typing import Any, Optional

from r2r.base import (
    Assistant,
    AsyncSyncMeta,
    GenerationConfig,
    Message,
    syncable,
)


class CombinedMeta(AsyncSyncMeta, ABCMeta):
    pass


class BaseR2RAssistant(Assistant, metaclass=CombinedMeta):

    @syncable
    async def arun(
        self,
        system_instruction: Optional[str] = None,
        messages: Optional[list[Message]] = None,
    ) -> str:
        messages_have_system_instruction = not (
            messages and "system" in messages
        )
        if system_instruction or messages_have_system_instruction:
            self._completed = False
            self._setup(system_instruction)

        if messages:
            for message in messages:
                print("message = ", message)
                self.conversation.add_message(message)

        while not self._completed:
            generation_config_with_functions = self.get_generation_config()
            result = await self.get_and_process_response(
                generation_config_with_functions
            )

            if self._completed:
                return result

    def get_generation_config(self) -> GenerationConfig:
        return GenerationConfig(
            **self.config.generation_config.model_dump(
                exclude={"functions", "stream"},
            ),
            functions=[
                {
                    "name": tool.name,
                    "description": tool.description,
                    "parameters": tool.parameters,
                }
                for tool in self.tools
            ],
            stream=self.is_streaming(),
        )

    @abstractmethod
    async def get_and_process_response(
        self, generation_config: GenerationConfig
    ) -> str:
        pass

    @abstractmethod
    def is_streaming(self) -> bool:
        pass

    async def handle_function_call(
        self, function_name: str, function_arguments: str
    ) -> str:
        tool_args = json.loads(function_arguments)
        self.conversation.create_and_add_message(
            "assistant",
            function_call={
                "name": function_name,
                "arguments": function_arguments,
            },
        )
        tool_result = await self.execute_tool(function_name, **tool_args)
        self.conversation.create_and_add_message(
            "function", content=tool_result, name=function_name
        )
        return await self.arun()  # Continue the conversation

    def handle_content_response(self, content: str) -> str:
        self.conversation.create_and_add_message("assistant", content=content)
        self._completed = True
        return content


class R2RAssistant(BaseR2RAssistant):
    def is_streaming(self) -> bool:
        return False

    async def get_and_process_response(
        self, generation_config: GenerationConfig
    ) -> str:
        response = await self.llm_provider.aget_completion(
            self.conversation.get_messages(),
            generation_config,
        )
        return await self.process_llm_response(response)

    async def process_llm_response(self, response: dict[str, Any]) -> str:
        message = response.choices[0].message
        if message.function_call:
            return await self.handle_function_call(
                message.function_call.name, message.function_call.arguments
            )
        else:
            return self.handle_content_response(message.content)


class R2RStreamingAssistant(BaseR2RAssistant):
    def is_streaming(self) -> bool:
        return True

    async def get_and_process_response(
        self, generation_config: GenerationConfig
    ) -> str:
        stream = self.llm_provider.get_completion_stream(
            self.conversation.get_messages(),
            generation_config,
        )
        return await self.process_llm_response(stream)

    async def process_llm_response(self, stream):
        function_name = None
        function_arguments = ""
        content = ""

        for chunk in stream:
            print("chunk = ", chunk)
            delta = chunk.choices[0].delta

            if delta.function_call:
                if delta.function_call.name:
                    function_name = delta.function_call.name
                if delta.function_call.arguments:
                    function_arguments += delta.function_call.arguments
            elif delta.content:
                content += delta.content

        if function_name:
            return await self.handle_function_call(
                function_name, function_arguments
            )
        else:
            return self.handle_content_response(content)
