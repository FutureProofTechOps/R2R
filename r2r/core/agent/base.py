"""Defines the abstract base classes and enums for agent objects"""

import logging
import logging.config
from abc import ABC, abstractmethod
from typing import Sequence

from ..abstractions.output import LLMChatCompletion
from ..abstractions.tool import Tool
from ..providers.llm import LLMProvider
from ..providers.prompt import PromptProvider

logger = logging.getLogger(__name__)


class Agent(ABC):
    """
    An abstract class for implementing a agent.

    An agent is an autonomous entity that can perform actions and communicate
    with other providers.
    """

    def __init__(
        self, prompt_provider: PromptProvider, llm_provider: LLMProvider
    ) -> None:
        self.llm_provider = llm_provider
        self.prompt_provider = prompt_provider

        self._initialized = False

    @abstractmethod
    def __iter__(self):
        pass

    @abstractmethod
    def __next__(self) -> LLMChatCompletion:
        """
        Iterates the agent by performing a single step of its task.

        A single step is a new conversation turn, which consists of generating
        a new 'asisstant' message, and parsing the reply from the 'user'.

        Raises:
            AgentStopIterationError: If the agent has already completed its task
            or exceeded the maximum number of iterations.
        """
        pass

    @property
    @abstractmethod
    def tools(self) -> Sequence[Tool]:
        """An abstract property for getting the tools associated with the agent."""
        pass

    @abstractmethod
    def run(self, message: str) -> str:
        """
        Runs the agent until it completes its task.

        The task is complete when next returns None.
        """
        pass

    @abstractmethod
    def run_stream(self, message: str):
        """
        Runs the agent until it completes its task. Stream the result when it is ready.
        """
        pass

    @abstractmethod
    def _setup(self) -> None:
        """An abstract method for setting up the agent before running."""
        pass
