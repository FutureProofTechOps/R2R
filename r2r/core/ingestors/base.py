import json
import string
from abc import ABC, abstractmethod
from io import BytesIO
from typing import Any, Generator, Generic, TypeVar, Union

from bs4 import BeautifulSoup

T = TypeVar("T")


class Ingestor(ABC, Generic[T]):
    @abstractmethod
    def ingest(self, data: Any) -> Generator[str, None, None]:
        pass


class TextAdapter(Ingestor[str]):
    def ingest(self, data: str) -> Generator[str, None, None]:
        yield data


class JSONAdapter(Ingestor[str]):
    def ingest(self, data: Union[str, bytes]) -> Generator[str, None, None]:
        if isinstance(data, bytes):
            data = data.decode("utf-8")
        yield self._parse_json(json.loads(data))

    def _parse_json(self, data: dict) -> str:
        def remove_objects_with_null(obj):
            if not isinstance(obj, dict):
                return obj
            result = obj.copy()
            for key, value in obj.items():
                if isinstance(value, dict):
                    result[key] = remove_objects_with_null(value)
                elif value is None:
                    del result[key]
            return result

        def format_json_as_text(obj, indent=0):
            lines = []
            indent_str = " " * indent

            if isinstance(obj, dict):
                for key, value in obj.items():
                    if isinstance(value, (dict, list)):
                        nested = format_json_as_text(value, indent + 2)
                        lines.append(f"{indent_str}{key}:\n{nested}")
                    else:
                        lines.append(f"{indent_str}{key}: {value}")
            elif isinstance(obj, list):
                for item in obj:
                    nested = format_json_as_text(item, indent + 2)
                    lines.append(f"{nested}")
            else:
                return f"{indent_str}{obj}"

            return "\n".join(lines)

        return format_json_as_text(remove_objects_with_null(data))


class HTMLAdapter(Ingestor[str]):
    def ingest(self, data: str) -> list[str]:
        soup = BeautifulSoup(data, "html.parser")
        yield soup.get_text()


class PDFAdapter(Ingestor[str]):
    def __init__(self):
        try:
            from pypdf import PdfReader

            self.PdfReader = PdfReader
        except ImportError:
            raise ValueError(
                "Error, `pypdf` is required to run `PyPDFAdapter`. Please install it using `pip install pypdf`."
            )

    def ingest(self, data: bytes) -> Generator[str, None, None]:
        pdf = self.PdfReader(BytesIO(data))
        for page in pdf.pages:
            page_text = page.extract_text()
            if page_text is not None:
                page_text = "".join(
                    filter(lambda x: x in string.printable, page_text)
                )
                yield page_text


class PPTAdapter(Ingestor[str]):
    def __init__(self):
        try:
            from pptx import Presentation

            self.Presentation = Presentation
        except ImportError:
            raise ValueError(
                "Error, `python-pptx` is required to run `PPTAdapter`. Please install it using `pip install python-pptx`."
            )

    def ingest(self, data: bytes) -> list[str]:
        prs = self.Presentation(BytesIO(data))
        for slide in prs.slides:
            for shape in slide.shapes:
                if hasattr(shape, "text"):
                    yield shape.text


class DOCXAdapter(Ingestor[str]):
    def __init__(self):
        try:
            from docx import Document

            self.Document = Document
        except ImportError:
            raise ValueError(
                "Error, `python-docx` is required to run `DOCXAdapter`. Please install it using `pip install python-docx`."
            )

    def ingest(self, data: bytes) -> Generator[str, None, None]:
        doc = self.Document(BytesIO(data))
        for paragraph in doc.paragraphs:
            yield paragraph.text


class CSVAdapter(Ingestor[str]):
    def __init__(self):
        import csv
        from io import StringIO

        self.csv = csv
        self.StringIO = StringIO

    def ingest(self, data: Union[str, bytes]) -> Generator[str, None, None]:
        if isinstance(data, bytes):
            data = data.decode("utf-8")
        csv_reader = self.csv.reader(self.StringIO(data))
        for row in csv_reader:
            yield ", ".join(row)


class XLSXAdapter(Ingestor[str]):
    def __init__(self):
        try:
            from openpyxl import load_workbook

            self.load_workbook = load_workbook
        except ImportError:
            raise ValueError(
                "Error, `openpyxl` is required to run `XLSXAdapter`. Please install it using `pip install openpyxl`."
            )

    def ingest(self, data: bytes) -> Generator[str, None, None]:
        wb = self.load_workbook(filename=BytesIO(data))
        for sheet in wb.worksheets:
            for row in sheet.iter_rows(values_only=True):
                yield ", ".join(map(str, row))


class MarkdownAdapter(Ingestor[str]):
    def __init__(self):
        import markdown

        self.markdown = markdown

    def ingest(self, data: str) -> Generator[str, None, None]:
        html = self.markdown.markdown(data)
        soup = BeautifulSoup(html, "html.parser")
        yield soup.get_text()
