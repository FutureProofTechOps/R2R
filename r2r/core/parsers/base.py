import json
import string
from abc import ABC, abstractmethod
from io import BytesIO
from typing import Any, AsyncGenerator, Generic, TypeVar, Union

from bs4 import BeautifulSoup

from ..abstractions.document import DataType

T = TypeVar("T")


class Parser(ABC, Generic[T]):
    @abstractmethod
    async def ingest(self, data: Any) -> AsyncGenerator[DataType, None]:
        pass


class TextParser(Parser[str]):
    async def ingest(self, data: str) -> AsyncGenerator[DataType, None]:
        yield data


class JSONParser(Parser[str]):
    async def ingest(
        self, data: Union[str, bytes]
    ) -> AsyncGenerator[str, None]:
        if isinstance(data, bytes):
            data = data.decode("utf-8")
        yield self._parse_json(json.loads(data))

    def _parse_json(self, data: dict) -> str:
        def remove_objects_with_null(obj):
            if not isinstance(obj, dict):
                return obj
            result = obj.copy()
            for key, value in obj.items():
                if isinstance(value, dict):
                    result[key] = remove_objects_with_null(value)
                elif value is None:
                    del result[key]
            return result

        def format_json_as_text(obj, indent=0):
            lines = []
            indent_str = " " * indent

            if isinstance(obj, dict):
                for key, value in obj.items():
                    if isinstance(value, (dict, list)):
                        nested = format_json_as_text(value, indent + 2)
                        lines.append(f"{indent_str}{key}:\n{nested}")
                    else:
                        lines.append(f"{indent_str}{key}: {value}")
            elif isinstance(obj, list):
                for item in obj:
                    nested = format_json_as_text(item, indent + 2)
                    lines.append(f"{nested}")
            else:
                return f"{indent_str}{obj}"

            return "\n".join(lines)

        return format_json_as_text(remove_objects_with_null(data))


class HTMLParser(Parser[str]):
    async def ingest(self, data: str) -> AsyncGenerator[str, None]:
        soup = BeautifulSoup(data, "html.parser")
        yield soup.get_text()


class PDFParser(Parser[str]):
    def __init__(self):
        try:
            from pypdf import PdfReader

            self.PdfReader = PdfReader
        except ImportError:
            raise ValueError(
                "Error, `pypdf` is required to run `PyPDFParser`. Please install it using `pip install pypdf`."
            )

    async def ingest(self, data: bytes) -> AsyncGenerator[str, None]:
        pdf = self.PdfReader(BytesIO(data))
        for page in pdf.pages:
            page_text = page.extract_text()
            if page_text is not None:
                page_text = "".join(
                    filter(lambda x: x in string.printable, page_text)
                )
                yield page_text


class PPTParser(Parser[str]):
    def __init__(self):
        try:
            from pptx import Presentation

            self.Presentation = Presentation
        except ImportError:
            raise ValueError(
                "Error, `python-pptx` is required to run `PPTParser`. Please install it using `pip install python-pptx`."
            )

    async def ingest(self, data: bytes) -> AsyncGenerator[str, None]:
        prs = self.Presentation(BytesIO(data))
        for slide in prs.slides:
            for shape in slide.shapes:
                if hasattr(shape, "text"):
                    yield shape.text


class DOCXParser(Parser[str]):
    def __init__(self):
        try:
            from docx import Document

            self.Document = Document
        except ImportError:
            raise ValueError(
                "Error, `python-docx` is required to run `DOCXParser`. Please install it using `pip install python-docx`."
            )

    async def ingest(self, data: bytes) -> AsyncGenerator[str, None]:
        doc = self.Document(BytesIO(data))
        for paragraph in doc.paragraphs:
            yield paragraph.text


class CSVParser(Parser[str]):
    def __init__(self):
        import csv
        from io import StringIO

        self.csv = csv
        self.StringIO = StringIO

    async def ingest(
        self, data: Union[str, bytes]
    ) -> AsyncGenerator[str, None]:
        if isinstance(data, bytes):
            data = data.decode("utf-8")
        csv_reader = self.csv.reader(self.StringIO(data))
        for row in csv_reader:
            yield ", ".join(row)


class XLSXParser(Parser[str]):
    def __init__(self):
        try:
            from openpyxl import load_workbook

            self.load_workbook = load_workbook
        except ImportError:
            raise ValueError(
                "Error, `openpyxl` is required to run `XLSXParser`. Please install it using `pip install openpyxl`."
            )

    async def ingest(self, data: bytes) -> AsyncGenerator[str, None]:
        wb = self.load_workbook(filename=BytesIO(data))
        for sheet in wb.worksheets:
            for row in sheet.iter_rows(values_only=True):
                yield ", ".join(map(str, row))


class MarkdownParser(Parser[str]):
    def __init__(self):
        import markdown

        self.markdown = markdown

    async def ingest(self, data: str) -> AsyncGenerator[str, None]:
        html = self.markdown.markdown(data)
        soup = BeautifulSoup(html, "html.parser")
        yield soup.get_text()
