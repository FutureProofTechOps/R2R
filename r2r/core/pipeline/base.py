import asyncio
import inspect
import uuid
from typing import Any, Optional

from ..pipes.base import AsyncPipe, AsyncState, PipeFlow
from ..pipes.logging import PipeLoggingConnectionSingleton
from ..utils import generate_run_id


class Pipeline:
    def __init__(
        self, pipe_logger: Optional[PipeLoggingConnectionSingleton] = None
    ):
        self.pipes: list[AsyncPipe] = []
        self.upstream_outputs: list[list[dict[str, str]]] = []
        self.pipe_logger = pipe_logger or PipeLoggingConnectionSingleton()
        self.futures = {}
        self.level = 0

    def add_pipe(
        self,
        pipe: AsyncPipe,
        add_upstream_outputs: Optional[list[dict[str, str]]] = None,
        *args,
        **kwargs,
    ) -> None:
        self.futures[pipe.config.name] = asyncio.Future()
        self.pipes.append(pipe)
        if not add_upstream_outputs:
            add_upstream_outputs = []
        self.upstream_outputs.append(add_upstream_outputs)

    async def run(self, input: Any, state: Optional[AsyncState] = None):
        self.state = state or AsyncState()
        current_input = input
        run_id = generate_run_id()
        print(f"Running pipeline with run_id: {run_id}")
        await self.pipe_logger.log(
            pipe_run_id=run_id,
            key="pipeline_type",
            value="rag",
            is_pipeline_info=True,
        )

        for pipe_num in range(len(self.pipes)):
            if self.pipes[pipe_num].flow == PipeFlow.FAN_OUT:
                if self.level == 0:
                    current_input = await self._run_pipe(
                        pipe_num, current_input, run_id
                    )
                    self.level += 1
                elif self.level == 1:
                    raise ValueError("Fan out not supported at level 1")
            elif self.pipes[pipe_num].flow == PipeFlow.STANDARD:
                if self.level == 0:
                    current_input = await self._run_pipe(
                        pipe_num, current_input, run_id
                    )
                elif self.level == 1:
                    current_input = [
                        await self._run_pipe(pipe_num, item)
                        async for item in current_input
                    ]
            elif self.pipes[pipe_num].flow == PipeFlow.FAN_IN:
                if self.level == 0:
                    raise ValueError("Fan in not supported at level 0")
                if self.level == 1:
                    current_input = await self._run_pipe(
                        pipe_num, current_input, run_id
                    )
                    self.level -= 1
            self.futures[self.pipes[pipe_num].config.name].set_result(
                current_input
            )

        # Consume the final generator to return a final result
        final_result = []
        async for item in current_input:
            final_result.append(item)
        return final_result if len(final_result) > 1 else final_result[0]

    async def _run_pipe(self, pipe_num: int, input: Any, run_id: uuid.UUID):
        # Collect inputs, waiting for the necessary futures
        pipe = self.pipes[pipe_num]
        add_upstream_outputs = self.upstream_outputs[pipe_num]
        input_dict = {"message": input}

        for upstream_input in add_upstream_outputs:
            upstream_pipe_name = upstream_input["prev_pipe_name"]

            async def resolve_future_output(future):
                result = future.result()
                # consume the async generator
                return [item async for item in result]

            async def replay_items_as_async_gen(items):
                for item in items:
                    yield item

            temp_results = await resolve_future_output(
                self.futures[upstream_pipe_name]
            )
            if upstream_pipe_name == self.pipes[pipe_num - 1].config.name:
                input_dict["message"] = replay_items_as_async_gen(temp_results)

            outputs = await self.state.get(
                upstream_input["prev_pipe_name"], "output"
            )
            prev_output_field = upstream_input.get("prev_output_field", None)
            if not prev_output_field:
                raise ValueError(
                    "`prev_output_field` must be specified in the upstream_input"
                )
            input_dict[upstream_input["input_field"]] = outputs[
                upstream_input["prev_output_field"]
            ]

        # Handle the pipe generator
        async for ele in pipe.run(
            pipe.Input(**input_dict), self.state, run_id=run_id
        ):
            return ele
