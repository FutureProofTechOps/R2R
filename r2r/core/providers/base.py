from abc import ABC, abstractmethod, abstractproperty
from dataclasses import fields
from typing import List, Optional


class ProviderConfig(ABC):
    """A base provider configuration class"""

    @abstractmethod
    def validate(self) -> None:
        pass

    @classmethod
    def create(cls, **kwargs):
        valid_keys = {f.name for f in fields(cls)}
        filtered_kwargs = {k: v for k, v in kwargs.items() if k in valid_keys}
        return cls(**filtered_kwargs)

    @abstractproperty
    def supported_providers(self) -> List[str]:
        """Define a list of supported providers."""
        pass


class Provider(ABC):
    """A base provider class to provide a common interface for all providers."""

    def __init__(self, config: Optional[ProviderConfig] = None):
        if config:
            config.validate()
        self.config = config
