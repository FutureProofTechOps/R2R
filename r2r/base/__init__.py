from .abstractions import *
from .agent import *
from .api.models import *
from .logging import *
from .parsers import *
from .pipeline import *
from .pipes import *
from .providers import *
from .utils import *

__all__ = [
    ## ABSTRACTIONS
    # Base abstractions
    "AsyncSyncMeta",
    "syncable",
    # Completion abstractions
    "CompletionRecord",
    "MessageType",
    # Document abstractions
    "DataType",
    "Document",
    "DocumentExtraction",
    "DocumentFragment",
    "DocumentInfo",
    "DocumentStatus",
    "DocumentType",
    # Embedding abstractions
    "EmbeddingPurpose",
    "default_embedding_prefixes",
    # Exception abstractions
    "R2RDocumentProcessingError",
    "R2RException",
    # KG abstractions
    "Entity",
    "KGExtraction",
    "Triple",
    "extract_triples",
    # Llama abstractions
    "VectorStoreQuery",
    # LLM abstractions
    "GenerationConfig",
    "LLMChatCompletion",
    "LLMChatCompletionChunk",
    "RAGCompletion",
    # Prompt abstractions
    "Prompt",
    # Search abstractions
    "AggregateSearchResult",
    "KGSearchResult",
    "KGSearchSettings",
    "VectorSearchResult",
    "VectorSearchSettings",
    # User abstractions
    "Token",
    "TokenData",
    "UserStats",
    # Vector abstractions
    "Vector",
    "VectorEntry",
    "VectorType",
    ## AGENT
    # Agent abstractions
    "Agent",
    "AgentConfig",
    "Conversation",
    "Message",
    "Tool",
    "ToolResult",
    ## API
    # Auth Requests
    "CreateUserRequest",
    "DeleteUserRequest",
    "LoginRequest",
    "LogoutRequest",
    "PasswordChangeRequest",
    "PasswordResetConfirmRequest",
    "PasswordResetRequest",
    "RefreshTokenRequest",
    "UserPutRequest",
    "VerifyEmailRequest",
    # Auth Responses
    "GenericMessageResponse",
    "TokenResponse",
    "UserResponse",
    # Ingestion Requests
    "R2RUpdateFilesRequest",
    "R2RIngestFilesRequest",
    # Management Requests
    "R2RUpdatePromptRequest",
    "R2RDeleteRequest",
    "R2RAnalyticsRequest",
    "R2RUsersOverviewRequest",
    "R2RDocumentsOverviewRequest",
    "R2RDocumentChunksRequest",
    "R2RLogsRequest",
    "R2RPrintRelationshipsRequest",
    "R2RCreateGroupRequest",
    "R2RUpdateGroupRequest",
    "R2RAddUserToGroupRequest",
    "R2RRemoveUserFromGroupRequest",
    "R2RGroupsOverviewRequest",
    "R2RScoreCompletionRequest",
    "R2RAssignDocumentToGroupRequest",
    "R2RRemoveDocumentFromGroupRequest",
    # Retrieval Requests
    "R2RSearchRequest",
    "R2RRAGRequest",
    "R2RAgentRequest",
    ## LOGGING
    # Basic types
    "RunType",
    "AnalysisTypes",
    "LogAnalytics",
    "LogAnalyticsConfig",
    "LogFilterCriteria",
    "LogProcessor",
    # Logging Providers
    "LocalRunLoggingProvider",
    "LoggingConfig",
    "PostgresLoggingConfig",
    "PostgresRunLoggingProvider",
    "RedisLoggingConfig",
    "RedisRunLoggingProvider",
    "RunLoggingSingleton",
    # Run Manager
    "RunManager",
    "manage_run",
    ## PARSERS
    # Base parser
    "AsyncParser",
    ## PIPELINE
    # Base pipeline
    "AsyncPipeline",
    ## PIPES
    "AsyncPipe",
    "AsyncState",
    "PipeType",
    ## PROVIDERS
    # Base provider classes
    "Provider",
    "ProviderConfig",
    # Auth provider
    "AuthConfig",
    "AuthProvider",
    # Chunking provider
    "ChunkingConfig",
    "ChunkingProvider",
    "Method",
    # Crypto provider
    "CryptoConfig",
    "CryptoProvider",
    # Database providers
    "DatabaseConfig",
    "DatabaseProvider",
    "RelationalDBProvider",
    "VectorDBProvider",
    "VectorDBFilterValue",
    # Embedding provider
    "EmbeddingConfig",
    "EmbeddingProvider",
    # Knowledge Graph provider
    "KGConfig",
    "KGDBProvider",
    "update_kg_prompt",
    "extract_entities",
    # LLM provider
    "CompletionConfig",
    "CompletionProvider",
    # Parsing provider
    "ParsingConfig",
    "ParsingProvider",
    "OverrideParser",
    # Prompt provider
    "PromptConfig",
    "PromptProvider",
    ## UTILS
    "RecursiveCharacterTextSplitter",
    "TextSplitter",
    "run_pipeline",
    "to_async_generator",
    "generate_run_id",
    "generate_id_from_label",
    "increment_version",
    "EntityType",
    "Relation",
    "format_entity_types",
    "format_relations",
]
