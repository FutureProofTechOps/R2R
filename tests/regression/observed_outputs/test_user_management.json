{
  "register_user": {
    "results": {
      "email": "test_e2ca9c68@example.com",
      "id": "a610b98f-99e4-5fff-a6ff-eea7e4806163",
      "group_ids": [],
      "hashed_password": "$2b$12$84yo.ii2U.tA3ceVhLC35.OeewG/Kd6hig6z.Yl/Yn8nDAL6gNyxW",
      "is_superuser": false,
      "is_active": true,
      "is_verified": false,
      "verification_code_expiry": null,
      "name": null,
      "bio": null,
      "profile_picture": null,
      "created_at": "2024-08-10T03:01:19.663863Z",
      "updated_at": "2024-08-10T03:01:19.663863Z"
    }
  },
  "login_user": {
    "results": {
      "access_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0X2UyY2E5YzY4QGV4YW1wbGUuY29tIiwiZXhwIjoxNzIzMjYyNDgwLjMyNTE5NiwidG9rZW5fdHlwZSI6ImFjY2VzcyJ9.3JOKTUcNBJXMZEaBybeM459xcHkqV7KWF_IZcaaiJdM",
        "token_type": "access"
      },
      "refresh_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0X2UyY2E5YzY4QGV4YW1wbGUuY29tIiwiZXhwIjoxNzIzODYzNjgwLCJ0b2tlbl90eXBlIjoicmVmcmVzaCJ9.C1d_wHgLZ6Udc3QsC8TE9U6TjvnRvwJbp56zDRrM3O8",
        "token_type": "refresh"
      }
    }
  },
  "user_info": {
    "results": {
      "email": "test_e2ca9c68@example.com",
      "id": "a610b98f-99e4-5fff-a6ff-eea7e4806163",
      "group_ids": [],
      "hashed_password": "$2b$12$84yo.ii2U.tA3ceVhLC35.OeewG/Kd6hig6z.Yl/Yn8nDAL6gNyxW",
      "is_superuser": false,
      "is_active": true,
      "is_verified": true,
      "verification_code_expiry": null,
      "name": null,
      "bio": null,
      "profile_picture": null,
      "created_at": "2024-08-10T03:01:19.663863Z",
      "updated_at": "2024-08-10T03:01:19.663863Z"
    }
  },
  "change_password": {
    "results": {
      "message": "Password changed successfully"
    }
  },
  "update_profile": {
    "results": {
      "email": "test_e2ca9c68@example.com",
      "id": "a610b98f-99e4-5fff-a6ff-eea7e4806163",
      "group_ids": [],
      "hashed_password": "null",
      "is_superuser": false,
      "is_active": true,
      "is_verified": true,
      "verification_code_expiry": null,
      "name": "John Doe",
      "bio": "R2R enthusiast",
      "profile_picture": null,
      "created_at": "2024-08-10T03:01:19.663863Z",
      "updated_at": "2024-08-10T03:01:21.374334Z"
    }
  },
  "refresh_token": {
    "results": {
      "access_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0X2UyY2E5YzY4QGV4YW1wbGUuY29tIiwiZXhwIjoxNzIzMjYyNDgxLjg3NDc1NSwidG9rZW5fdHlwZSI6ImFjY2VzcyJ9.PoIHwL1njH-xAvMzd1XS9OlmXsjIxYsAqE3MkLXqvMU",
        "token_type": "access"
      },
      "refresh_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0X2UyY2E5YzY4QGV4YW1wbGUuY29tIiwiZXhwIjoxNzIzODYzNjgxLCJ0b2tlbl90eXBlIjoicmVmcmVzaCJ9.TLMR6ZJPf3fdFfnPjsn_NaD_3Iv81lJ0S6N7YuT8b6U",
        "token_type": "refresh"
      }
    }
  },
  "superuser_test": {
    "results": [
      {
        "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
        "num_files": 14,
        "total_size_in_bytes": 4876790,
        "document_ids": [
          "cafdf784-a1dc-5103-8098-5b0a97db1707",
          "e0fc8bbc-95be-5a98-891f-c17a43fa2c3d",
          "bca7fd1c-bb71-5652-9734-620af8e3aafc",
          "e0c2de57-171d-5385-8081-b546a2c63ce3",
          "7f39e670-a61b-5506-a622-1638b4c62e45",
          "ad0f4c78-8e84-5201-8f35-fa83b3fcef06",
          "f17eac52-a22e-5c75-af8f-0b25b82d43f8",
          "022fdff4-f87d-5b0c-82e4-95d53bcc4e60",
          "c5b31b3a-06d2-553e-ac3e-47c56139b484",
          "97230d1c-33ae-5a9d-b93c-5c42dde68002",
          "b21a46a4-2906-5550-9529-087697da2944",
          "79145cf2-123a-5333-8bf4-ded6dde616fd",
          "6d5f8844-2da0-5558-8c17-b1d80906c175",
          "93123a68-d668-51de-8291-92162730dc87"
        ]
      }
    ]
  },
  "logout": {
    "results": {
      "message": "Logged out successfully"
    }
  },
  "delete_account": {
    "results": "Incorrect password"
  }
}
