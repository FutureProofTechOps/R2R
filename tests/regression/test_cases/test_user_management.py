import uuid

from tests.regression.test_cases.base import BaseTest


class TestUserManagement(BaseTest):
    def __init__(self, client):
        super().__init__(client)
        self.set_exclude_paths(
            "register_user",
            [
                f"root['results']['{field}']"
                for field in [
                    "id",
                    "email",
                    "created_at",
                    "updated_at",
                    "hashed_password",
                ]
            ],
        )
        self.set_exclude_paths(
            "login_user",
            [
                f"root['results']['{field}']"
                for field in ["access_token", "refresh_token"]
            ],
        )
        self.set_exclude_paths("user_info", [])

    def get_test_cases(self):
        return {
            "register_user": lambda client: self.register_user_test(client),
            "login_user": lambda client: self.login_user_test(client),
            "user_info": lambda client: self.user_info_test(client),
        }

    def register_user_test(self, client):
        try:
            email = f"test@example.com"
            password = "password123"
            return client.register(email, password)
        except:
            pass

        email = f"test_{uuid.uuid4()}@example.com"
        password = "password123"
        return client.register(email, password)

    def login_user_test(self, client):
        email = "test@example.com"  # Use a known test user
        password = "password123"
        return client.login(email, password)

    def user_info_test(self, client):
        return client.user()
