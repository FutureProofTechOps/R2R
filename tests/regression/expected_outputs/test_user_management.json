{
  "register_user": {
    "results": {
      "email": "test_a55d8044-d83a-424c-8d11-3d4131bf129c@example.com",
      "id": "791e50a3-5c74-5d3f-b138-927c86832f46",
      "group_ids": [],
      "hashed_password": "$2b$12$DvBP/kOaWCJJwaGQaGSubufX7rEwv311eCGdMDcYlqUTRnBFG.Dfm",
      "is_superuser": false,
      "is_active": true,
      "is_verified": false,
      "verification_code_expiry": null,
      "name": null,
      "bio": null,
      "profile_picture": null,
      "created_at": "2024-08-09T21:36:15.111262Z",
      "updated_at": "2024-08-09T21:36:15.111262Z"
    }
  },
  "login_user": {
    "results": {
      "access_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0QGV4YW1wbGUuY29tIiwiZXhwIjoxNzIzMjQyOTc1LjU3NzIwOSwidG9rZW5fdHlwZSI6ImFjY2VzcyJ9.8iuBDzjxg_1H-Te6kUP3Ut56qf7LB0JuBF0DmOBwoRI",
        "token_type": "access"
      },
      "refresh_token": {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0QGV4YW1wbGUuY29tIiwiZXhwIjoxNzIzODQ0MTc1LCJ0b2tlbl90eXBlIjoicmVmcmVzaCJ9.qS2-0Tqk0I0PdLbLmr3UKCLhjdC57xOf87IYvK7JIYs",
        "token_type": "refresh"
      }
    }
  },
  "user_info": {
    "results": {
      "email": "test@example.com",
      "id": "fcbcbc64-f85c-5025-877c-37f4c7a12d6e",
      "group_ids": [],
      "hashed_password": "$2b$12$5zQct0/el3uLCmxZaWZ3B.56mzuChojd3V5OF892AYrfD7/Yopa.G",
      "is_superuser": false,
      "is_active": true,
      "is_verified": true,
      "verification_code_expiry": null,
      "name": null,
      "bio": null,
      "profile_picture": null,
      "created_at": "2024-08-09T21:23:11.863355Z",
      "updated_at": "2024-08-09T21:23:11.863355Z"
    }
  }
}
